vhost/src/vhost_user/gpu_backend_req.rs

```diff
@@ -41,6 +41,20 @@ impl BackendInternal {
         self.wait_for_ack(&hdr)
     }
 
+    fn send_message_no_reply<T: ByteValued>(
```

Comment:
Why not just adding a parameter to `send_message()`.

That said, if `VHOST_USER_PROTOCOL_F_REPLY_ACK` is negotiated, usually all the messages should expect an ACK, is this an exception in vhost-user-gpu? Maybe we should document this.

Reply:
There is no `VHOST_USER_PROTOCOL_F_REPLY_ACK` flag in this protocol, so we don't expect acks for every message. 
Having a parameter for ack, would be messy I think, because this method has an generic argument for the type of reply, and the type of reply has to logically be `()` when not waiting for reply. Instead I think it could be achieved by having a special return type for no-ack messages with some trait. 

Another alternative could be calling `wait_for_ack` in the the public command methods depending if the command waits or doesn't wait.

Reply:
> Another alternative could be calling wait_for_ack in the the public command methods depending if the command waits or doesn't wait.

Yeah, I like this. 
So IIUC, some messages need a reply, so maybe instead of `wait_for_ack` we can call it `recv_reply`.

So the API will be:
- `send_header()`
- `send_message()`
- `send_message_with_payload()`
- `recv_reply()`

And call `recv_reply()` where needed. (I'd remove it also from `send_header()`).

IMHO this makes the code a little more understandable and reviewable, going to check for each message when a reply is required or not.

Reply:
After changing it to this. Do you think it is still enough to have tests just for this internal API (`send_header()`, `send_message()`, `send_message_with_payload()`, `recv_reply()`) ? Or should I change the tests to test the all of the public methods instead of internal API?

Reply:
> Hmm, actually I just realized there could be a theoretical issue with this. Because unlocking the mutex between the calls to `send_message()` and `recv_reply()` could theoretically not be correct, depending on the frontend implementation. Because in between these calls another `send_message()` could be issued and we could receive a reply for that one. As far as I can tell, this cannot happen in current QEMU implementation though, because messages are not processed in parallel. If we don't care about this I wonder if we even need to wrap this in a mutex.

Does the specification cover this case?

I thought in general we don't support sending two requests without waiting for a response to the first one first, in any case I think your approach is good.


> Doing this with the mutex would mean doing something like this which I guess is not that bad:

Yep, I think this is fine.